File: pdftract-dotnet/src/Pdftract/Models/Page.cs
using MessagePack;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Pdftract.Models;

/// <summary>
/// Represents a single page in the document.
/// </summary>
[MessagePackObject]
public record Page
{
    [Key(0)]
    [JsonPropertyName("page_index")]
    public required uint PageIndex { get; init; }

    [Key(1)]
    [JsonPropertyName("page_number")]
    public required uint PageNumber { get; init; }

    [Key(2)]
    [JsonPropertyName("page_label")]
    public string? PageLabel { get; init; }

    [Key(3)]
    [JsonPropertyName("width")]
    public required float Width { get; init; }

    [Key(4)]
    [JsonPropertyName("height")]
    public required float Height { get; init; }

    [Key(5)]
    [JsonPropertyName("rotation")]
    public required uint Rotation { get; init; }

    [Key(6)]
    [JsonPropertyName("type")]
    public required string PageType { get; init; }

    [Key(7)]
    [JsonPropertyName("spans")]
    public IList<Span> Spans { get; init; } = new List<Span>();

    [Key(8)]
    [JsonPropertyName("blocks")]
    public IList<Block> Blocks { get; init; } = new List<Block>();

    [Key(9)]
    [JsonPropertyName("tables")]
    public IList<Table> Tables { get; init; } = new List<Table>();

    [Key(10)]
    [JsonPropertyName("annotations")]
    public IList<Annotation> Annotations { get; init; } = new List<Annotation>();
}

File: pdftract-java/src/main/java/com/jedarden/pdftract/Page.java
package com.jedarden.pdftract;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;

/**
 * A single page in the document.
 */
public record Page(
    @JsonProperty("page_index") int pageIndex,
    @JsonProperty("width") double width,
    @JsonProperty("height") double height,
    @JsonProperty("rotation") int rotation,
    @JsonProperty("page_type") String pageType,
    @JsonProperty("spans") List<Span> spans,
    @JsonProperty("blocks") List<Block> blocks
) {
    public Page {
        spans = spans != null ? spans : List.of();
        blocks = blocks != null ? blocks : List.of();
    }
}
